lesson-05-time-management/usrmod.c

```diff
@@ -0,0 +1,108 @@
+#include <linux/module.h>
+#include <linux/fs.h>
+#include <linux/cdev.h>
+#include <asm/uaccess.h>
+#include <linux/pci.h>
+#include <linux/version.h>
+#include <linux/init.h>
+#include <linux/timer.h>
+#include <linux/jiffies.h>
+#include <linux/kernel.h>
+
+#include "fib.c"
+
+static struct timer_list one_sec_tmr;
+
+static char* out_buf;
+
+fb_t fb_num;
+
+static struct class *f_class;
+
+static uint32_t lst_reading_tmr = 0;
+
+static ssize_t feb_show( struct class *class, struct class_attribute *attr, char *buf ) {
+   fib_get_str(fb_num.fib_buf1, out_buf, fb_num.fib_curr_len);
+   strcpy( buf, out_buf );
+   printk( "read %ld bytes [Fibonacci #%u]\n", (long)strlen( buf ), fb_num.fib_iter );
+   return strlen( buf );
+}
+
+static ssize_t lst_r_tmr_show( struct class *class, struct class_attribute *attr, char *buf ) {
+    if(!lst_reading_tmr)
+    {
+        const char out[] = "Newer\n";
+        strcpy(buf, out);
+        
+    }
+    else
+    {
+        char out[50];
+        snprintf(out, 50, "last reading was %u secs ago\n", jiffies_to_msecs(jiffies)/1000-lst_reading_tmr);
+        strcpy(buf, out);
+    }
+    lst_reading_tmr = jiffies_to_msecs(jiffies)/1000;
+    return strlen( buf );
+}
+
+
+CLASS_ATTR_RO( feb );
+CLASS_ATTR_RO( lst_r_tmr );
+
+void timer_handler(unsigned long data)
+{
+    if(!fib_step(&fb_num))
```

Comment:
It's preferable to use data pointer instead of global variables to pass structure.